### PR TITLE
remote: track listeners on browser side

### LINKED ISSUE #3229: [IPC] Adding and removing listeners from a remote EventEmitter
I'm using the `auto-updater` module, which exports an EventEmitter instance. Because I want to easily display the update status in the UI, I import it with `require('remote').require('auto-updater')` in the renderer process.

The issue I've run into is that the remote module/IPC interface doesn't seem to track the identities of functions passed to it. That is, this code doesn't actually remove the listener:
```js
autoUpdater.on('checking-for-update', listener);
autoUpdater.removeListener('checking-for-update', listener);
```

So presumably the internals of `remote` or the IPC layer should save the listeners as keys of a Map/WeakMap that are mapped to callback IDs. When the same local function is passed to multiple remote method calls, the remote side should know that it's the same function.

### PR DESCRIPTION
**Pull request by deepak1556**
Fixes #3229

### CODE REVIEW

File: atom/browser/lib/rpc-server.coffee [hunk 1]
```diff
@@ -3,6 +3,9 @@ path = require 'path'
 objectsRegistry = require './objects-registry.js'
 v8Util = process.atomBinding 'v8_util'
 
+# weak refereence to callback with their registry ID.
```
Comment (zcbenz) [inline atom/browser/lib/rpc-server.coffee]: First character of comment line should be capitalized.

File: atom/browser/lib/rpc-server.coffee [hunk 2]
```diff
@@ -74,14 +77,19 @@ unwrapArgs = (sender, args) ->
         objectsRegistry.once "clear-#{sender.getId()}", ->
           rendererReleased = true
 
+        if rendererCallbacks.has(meta.id)
```
Comment (zcbenz) [inline atom/browser/lib/rpc-server.coffee]: It should be put earlier than the `rendererReleased` line.

**Comment by zcbenz** [comment]
The `atom::WeakMap` is an RAII class used by C++ classes, if you want to expose it to JavaScript you should add an binding to it instead of making it a `Wrappable` directly.

**Comment by deepak1556** [comment]
@zcbenz have made the changes.

**Comment by zcbenz** [comment]
Using `WeakMap` is going to conflict with ES6's `WeakMap`, I still prefer to use `IDWeakMap`, which also matches the C++ class's name.

File: atom/browser/lib/rpc-server.coffee [hunk 3]
```diff
@@ -70,6 +74,9 @@ unwrapArgs = (sender, args) ->
         returnValue = metaToValue meta.value
         -> returnValue
       when 'function'
+        if rendererCallbacks.has(meta.id)
```
Comment (zcbenz) [inline atom/browser/lib/rpc-server.coffee]: Different WebContents can send callbacks with the same ID.

Comment (deepak1556) [inline atom/browser/lib/rpc-server.coffee]: Fixed, thanks!

File: atom/browser/lib/rpc-server.coffee [hunk 4]
```diff
@@ -70,18 +74,30 @@ unwrapArgs = (sender, args) ->
         returnValue = metaToValue meta.value
         -> returnValue
       when 'function'
+        webContentsId = sender.getId()
+        rendererCallbacks = rendererRegistry[webContentsId]
+        if not rendererCallbacks?
+          # Weak reference to callbacks with their ID
+          rendererCallbacks = new IDWeakMap()
+          rendererRegistry[webContentsId] = rendererCallbacks
```
Comment (zcbenz) [inline atom/browser/lib/rpc-server.coffee]: The `rendererCallbacks` should be destroyed when the render view of webContents is destroyed.

File: atom/common/api/atom_api_id_weak_map.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  bool IsDestroyed() const override;
+
+ private:
+  // Api for IDWeakMap.
+  void Set(v8::Isolate* isolate, int32_t id, v8::Local<v8::Object> object);
+  v8::Local<v8::Object> Get(v8::Isolate* isolate, int32_t id);
+  bool Has(int32_t id);
+  void Remove(int32_t id);
+
+  scoped_ptr<atom::IDWeakMap> id_weak_map_;
```
Comment (zcbenz) [inline atom/common/api/atom_api_id_weak_map.h]: It doesn't seem necessary to use `scoped_ptr`.

**Comment by zcbenz** [comment]
Can you reimplement `IDWeakMap::Add` with `IDWeakMap::Set`? Their code is basically the same.

**Comment by deepak1556** [comment]
@zcbenz have addressed them, thanks!

**Comment by zcbenz** [comment]
Thanks.

**Pull request merged**